File: src/reporter.ts
/**
 * reporter.ts - ASCII 报告生成器
 *
 * 职责：将回测结果格式化为终端友好的 ASCII 文本报告。
 *
 * 输出内容：
 *   1. 概览表头
 *   2. 资金统计（含收益率、最大回撤）
 *   3. 交易统计（次数、胜率）
 *   4. 与买入持有对比
 *   5. 价格走势 + 买卖信号 ASCII 图
 */

import { BacktestResult } from './backtester.js';
import { StrategyCandle } from './strategy.js';

/** 百分比格式化 */
function pct(n: number): string {
  const sign = n >= 0 ? '+' : '';
  return `${sign}${(n * 100).toFixed(2)}%`;
}

/** 金额格式化 */
function usd(n: number): string {
  return '$' + n.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

/** 画分隔线 */
const SEP = '═'.repeat(50);

/**
 * 生成完整回测报告
 */
export function generateReport(
  candles: StrategyCandle[],
  result: BacktestResult,
  symbol: string,
  shortPeriod: number,
  longPeriod: number,
  interval: string
): string {
  const lines: string[] = [];

  // ==================== 表头 ====================
  lines.push(SEP);
  lines.push('  MA 交叉策略回测报告');
  lines.push(`  交易对: ${symbol.padEnd(14)} 周期: ${interval}`);
  lines.push(`  短均线: ${shortPeriod.toString().padEnd(14)} 长均线: ${longPeriod}`);
  lines.push(SEP);
  lines.push('');

  // ==================== 资金统计 ====================
  lines.push('  ── 资金统计 ──');
  lines.push(`  初始资金: ${usd(result.initialCapital)}`);
  lines.push(`  最终资金: ${usd(result.finalCapital)}`);
  lines.push(`  总收益率: ${pct(result.totalReturn)}`);
  lines.push(`  最大回撤: ${pct(-result.maxDrawdown)}`);
  lines.push('');

  // ==================== 交易统计 ====================
  const pairs = Math.floor(result.trades.length / 2);
  const winRate = pairs > 0 ? (result.winningTrades / pairs * 100) : 0;
  lines.push('  ── 交易统计 ──');
  lines.push(`  交易次数: ${result.totalTrades} (${pairs} 次完整买卖)`);
  lines.push(`  胜率:     ${winRate.toFixed(1)}%`);
  lines.push('');

  // ==================== 对比基准 ====================
  lines.push('  ── 基准对比 ──');
  lines.push(`  买入持有: ${pct(result.buyAndHoldReturn)}`);
  const excess = result.totalReturn - result.buyAndHoldReturn;
  lines.push(`  策略超额: ${pct(excess)}`);
  lines.push('');

  // ==================== 交易明细 ====================
  lines.push('  ── 交易明细 ──');
  if (result.trades.length === 0) {
    lines.push('  (无交易)');
  } else {
    for (let i = 0; i < result.trades.length; i++) {
      const t = result.trades[i];
      const date = new Date(t.time).toISOString().slice(0, 16).replace('T', ' ');
      const arrow = t.type === 'buy' ? '买 ↑' : '卖 ↓';
      lines.push(
        `  ${date}  ${arrow}  ${t.price.toFixed(2).padStart(10)}  × ${t.amount.toFixed(4).padStart(8)} = ${usd(t.value)}`
      );
    }
  }
  lines.push('');

  // ==================== ASCII 走势图 ====================
  lines.push('  ── 价格走势 ──');
  lines.push(drawChart(candles));
  lines.push('');

  lines.push(SEP);

  return lines.join('\n');
}

/**
 * 绘制简易 ASCII 价格走势图 + 买卖信号标记
 *
 * 原理：归一化价格到 [0, chartHeight]，用字符画折线。
 */
function drawChart(candles: StrategyCandle[]): string {
  // 采样：如果数据太多，降采样到 80 个点
  const MAX_POINTS = 80;
  const step = Math.max(1, Math.floor(candles.length / MAX_POINTS));
  const sampled: StrategyCandle[] = [];
  for (let i = 0; i < candles.length; i += step) {
    sampled.push(candles[i]);
  }

  const chartHeight = 8;
  const closes = sampled.map((c) => c.close);
  const min = Math.min(...closes);
  const max = Math.max(...closes);
  const range = max - min || 1;

  // 归一化价格到 0..chartHeight
  const norm = (price: number) =>
    Math.round(((price - min) / range) * chartHeight);

  const lines: string[] = [];

  // 从上到下画
  for (let row = chartHeight; row >= 0; row--) {
    let line = '  ';
    if (row === chartHeight) line += max.toFixed(0).padStart(6) + '┤';
    else if (row === Math.floor(chartHeight / 2)) line += ' '.repeat(6) + '┤';
    else if (row === 0) line += min.toFixed(0).padStart(6) + '┤';
    else line += ' '.repeat(7);

    for (let i = 0; i < sampled.length; i++) {
      const y = norm(sampled[i].close);
      const prevY = i > 0 ? norm(sampled[i - 1].close) : y;

      if (y === row) {
        // 信号标记优先于折线
        if (sampled[i].signal === 'buy') {
          line += 'B'; // Buy
        } else if (sampled[i].signal === 'sell') {
          line += 'S'; // Sell
        } else {
          line += '─';
        }
      } else if (prevY < row && y > row || prevY > row && y < row) {
        line += '│'; // 竖线连接
      } else {
        line += ' ';
      }
    }
    lines.push(line);
  }

  // X 轴
  lines.push('  ' + ' '.repeat(6) + '└' + '─'.repeat(sampled.length) + '→ 时间');
  lines.push('  ' + ' '.repeat(6) + ' B=买入  S=卖出');

  return lines.join('\n');
}

File: src/reporter.py
"""reporter.py — 富文本报告生成器

使用 Rich 输出终端友好的彩色报告。
"""
from datetime import datetime, timezone
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from rich.text import Text

from src.backtester import BacktestResult
from src.sentiment import get_sentiment_score

console = Console()


def _pct(v: float) -> str:
    color = "green" if v >= 0 else "red"
    return f"[{color}]{v*100:+.2f}%[/{color}]"


def _usd(v: float) -> str:
    return f"${v:,.2f}"


def print_price(symbol: str, price: float):
    """实时价格输出"""
    sent = get_sentiment_score()
    text = Text()
    text.append(f"\n  {symbol}  ", style="bold cyan")
    text.append(f"{_usd(price)}", style="bold yellow")
    text.append(f"\n  市场情绪: {sent['classification']} ({sent['raw_value']})", style="dim")
    console.print(text)


def print_report(
    df,
    result: BacktestResult,
    symbol: str,
    short: int,
    long: int,
    interval: str,
):
    """主报告"""
    console.print()

    # ── 概览面板 ──
    title = Text("MA 交叉策略回测报告", style="bold white on blue")
    overview = Table.grid(padding=(0, 2))
    overview.add_column(style="dim")
    overview.add_column()
    overview.add_row("交易对", f"[cyan]{symbol}[/cyan]")
    overview.add_row("周期", interval)
    overview.add_row("短均线", str(short))
    overview.add_row("长均线", str(long))
    overview.add_row("K 线数", str(len(df)))
    overview.add_row("数据范围", f"{_ts2str(df['time'].iloc[0])} ~ {_ts2str(df['time'].iloc[-1])}")

    # 当前情绪
    sent = get_sentiment_score()
    sent_color = "red" if sent["score"] < -0.3 else "green" if sent["score"] > 0.3 else "yellow"
    overview.add_row("当前情绪", f"[{sent_color}]{sent['classification']} ({sent['raw_value']})[/{sent_color}]")

    console.print(Panel(overview, title=title, border_style="blue"))
    console.print()

    # ── 收益统计 ──
    perf = Table(title="资金统计", border_style="green")
    perf.add_column("指标", style="dim")
    perf.add_column("数值", justify="right")
    perf.add_row("初始资金", _usd(result.initial_capital))
    perf.add_row("最终资金", _usd(result.final_capital))
    perf.add_row("总收益率", _pct(result.total_return))
    perf.add_row("最大回撤", _pct(-result.max_drawdown))
    perf.add_row("夏普比率", f"{result.sharpe_ratio:.2f}")
    perf.add_row("买入持有", _pct(result.buy_hold_return))
    excess = result.total_return - result.buy_hold_return
    perf.add_row("策略超额", _pct(excess))
    if result.sentiment_avg != 0:
        mood = "恐惧" if result.sentiment_avg < -0.2 else "贪婪" if result.sentiment_avg > 0.2 else "中性"
        perf.add_row("持仓情绪均值", f"{result.sentiment_avg:+.2f} ({mood})")
    console.print(perf)
    console.print()

    # ── 交易统计 ──
    buys = [t for t in result.trades if t.type == "buy"]
    sells = [t for t in result.trades if t.type == "sell"]
    pairs = min(len(buys), len(sells))
    wins = sum(
        1 for i in range(pairs)
        if sells[i].value > buys[i].value
    )
    win_rate = (wins / pairs * 100) if pairs > 0 else 0

    trade_table = Table(title="交易统计", border_style="yellow")
    trade_table.add_column("指标", style="dim")
    trade_table.add_column("数值", justify="right")
    trade_table.add_row("总交易笔数", str(len(result.trades)))
    trade_table.add_row("完整买卖次数", str(pairs))
    trade_table.add_row("胜率", f"[{'green' if win_rate > 50 else 'red'}]{win_rate:.1f}%[/]")
    avg_profit = sum(sells[i].value - buys[i].value for i in range(pairs)) / pairs if pairs else 0
    color = "green" if avg_profit > 0 else "red"
    trade_table.add_row("平均每笔盈亏", f"[{color}]{_usd(avg_profit)}[/{color}]")
    console.print(trade_table)
    console.print()

    # ── 交易明细（最近 10 笔） ──
    detail = Table(title="最近交易明细", border_style="dim")
    detail.add_column("时间", style="dim")
    detail.add_column("方向")
    detail.add_column("价格", justify="right")
    detail.add_column("金额", justify="right")
    for t in result.trades[-10:]:
        arrow = "[green]买 ↑[/green]" if t.type == "buy" else "[red]卖 ↓[/red]"
        detail.add_row(
            _ts2str(t.time),
            arrow,
            f"{t.price:,.2f}",
            _usd(t.value),
        )
    console.print(detail)
    console.print()


def print_weights(weights: dict):
    """输出信号权重表"""
    t = Table(title="信号权重配置", border_style="magenta")
    t.add_column("信号源", style="dim")
    t.add_column("权重", justify="right")
    for k, v in weights.items():
        t.add_row(k, f"{v:.0%}")
    console.print(t)
    console.print()


def _ts2str(ts: int) -> str:
    return datetime.fromtimestamp(ts / 1000, tz=timezone.utc).strftime("%Y-%m-%d %H:%M")
